compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaLexer.java

```diff
@@ -1550,4 +1567,296 @@ private STToken readTokenInBracedContentInInterpolation() {
+            }
+        }
+    }
+
+    private boolean processDocumentationReference(int nextChar) {
+        // Look ahead and see if next characters belong to a documentation reference.
+        // If they do, do not advance the reader and return.
+        // Otherwise advance the reader for checked characters and return
+
+        int lookAheadChar = nextChar;
+        int lookAheadCount = 0;
+        String identifier = "";
+
+        while (isIdentifierInitialChar(lookAheadChar)) {
+            identifier = identifier.concat(String.valueOf((char) lookAheadChar));
+            lookAheadCount++;
+            lookAheadChar = reader.peek(lookAheadCount);
+        }
+
+        switch (identifier) {
+            case LexerTerminals.TYPE:
+            case LexerTerminals.SERVICE:
+            case LexerTerminals.VARIABLE:
+            case LexerTerminals.VAR:
+            case LexerTerminals.ANNOTATION:
+            case LexerTerminals.MODULE:
+            case LexerTerminals.FUNCTION:
+            case LexerTerminals.PARAMETER:
+                // Look ahead for a single backtick.
+                // There could be spaces or tabs in between.
+                while (true) {
+                    switch (lookAheadChar) {
+                        case LexerTerminals.SPACE:
+                        case LexerTerminals.TAB:
+                            lookAheadCount++;
+                            lookAheadChar = reader.peek(lookAheadCount);
+                            continue;
+                        case LexerTerminals.BACKTICK:
+                            // Make sure backtick is a single backtick
+                            if (reader.peek(lookAheadCount + 1) != LexerTerminals.BACKTICK) {
+                                // Reaching here means checked characters belong to a documentation reference.
+                                // Hence return.
+                                return true;
+                            }
+                            // Fall through
+                        default:
+                            break;
+                    }
+                    break;
+                }
+                // Fall through
+            default:
+                //
+                reader.advance(lookAheadCount);
+                return false;
+        }
+    }
+
+    /*
+     * ------------------------------------------------------------------------------------------------------------
+     * DOCUMENTATION_PARAMETER Mode
+     * ------------------------------------------------------------------------------------------------------------
+     */
+
+    private STToken readDocumentationParameterToken() {
+        reader.mark();
+        int nextChar = peek();
+        if (isIdentifierInitialChar(nextChar)) {
+            STToken token;
+            reader.advance();
+            while (isIdentifierInitialChar(peek())) {
+                reader.advance();
+            }
+            if (getLexeme().equals("return")) {
```

Comment:
Shall we use the constant `LexerTerminals.RETURN`.

BTW, usually we swap the arguments as a practice to make the `.equals()` is called on the constant.. 
eg: `"return".equals(getLexeme())` or `LexerTerminals.RETURN.equals(getLexeme())`

That way we can avoid any possible NPE.  i.e: even if `getLexeme()` returns a null, it will not throw a NPE